### PR TITLE
Fix #503, better events for CFE_TBL_Load()

### LINKED ISSUE #503: CFE_TBL_Load should produce clearer message when it has a file header mismatch
**Is your feature request related to a problem? Please describe.**
CFE_TBL_LoadFromFile compares the application's name + table name with the string in the header of the file. If you configure the cfe_es_startup.scr with a different application name than what is expected in the header, you get an obtuse "Fail to load Tbl '<app>.<tbl>' from '/cf/<file>.tbl'" not making it clear that it's purely a mismatch of the header.

**Describe the solution you'd like**
At minimum it should indicate that the matter is a mis-match of the header and the expected header (which means that the file is there, for example). Better would be to report the actual contents of the header and the expected value. The EID for the error is unique so CFE_TBL_Load can produce an event with more specific text.

**Describe alternatives you've considered**
Could have CFE_TBL_LoadFromFile generate the event, but that's an "internal" fn so probably not.

**Additional context**
Add any other context about the feature request here.

**Requester Info**
Christopher.D.Knight@nasa.gov

### PR DESCRIPTION
**Pull request by CDKnightNASA**
**Describe the contribution**
Addresses #503 but this is a draft, still need UT code updates. Simplified changes for CFE_TBL_Load() to make the code a bit easier to follow (hey, events generated when results are off-nominal, instead of at the end?)

**Testing performed**
Initial build/run, UT not updated yet.

**Expected behavior changes**
Internal CFE_TBL_LoadFromFile() API changed slightly to add AppName as a parameter. Return value from LoadFromFile() no longer relevant for event generation.

**System(s) tested on**
Debian 9

**Contributor Info - All information REQUIRED for consideration of pull request**
Christopher.D.Knight@nasa.gov

### CODE REVIEW

**Comment by CDKnightNASA** [comment]
Updated the unit tests, removing "draft" status from pull request.

File: fsw/cfe-core/src/tbl/cfe_tbl_api.c [hunk 1]
```diff
@@ -692,261 +692,222 @@ int32 CFE_TBL_Load( CFE_TBL_Handle_t TblHandle,
...
-                                           CFE_EVS_EventType_DEBUG,
-                                           CFE_TBL_TaskData.TableTaskAppId,
-                                           "Successfully loaded '%s' from '%s'",
-                                           RegRecPtr->Name,
-                                           RegRecPtr->Buffers[0].DataSource);
-            }
-            else
-            {
-                Status = CFE_TBL_ERR_DUMP_ONLY;
+    /* Get pointers to pertinent records in registry and handles */
```
Comment (skliper) [inline fsw/cfe-core/src/tbl/cfe_tbl_api.c:718]: Stray comment?

File: fsw/cfe-core/src/tbl/cfe_tbl_api.c [hunk 2]
```diff
@@ -692,261 +692,222 @@ int32 CFE_TBL_Load( CFE_TBL_Handle_t TblHandle,
...
+    CFE_ES_GetAppName(AppName, ThisAppId, OS_MAX_API_NAME);
 
     /* Initialize return pointer to NULL */
     WorkingBufferPtr = NULL;
 
     /* Verify access rights and get a valid Application ID for calling App */
     Status = CFE_TBL_ValidateAccess(TblHandle, &ThisAppId);
 
-    if (Status == CFE_SUCCESS)
+    if (Status != CFE_SUCCESS)
```
Comment (skliper) [inline fsw/cfe-core/src/tbl/cfe_tbl_api.c:709]: This routine mixes use of status and returns throughout that make's it hard to follow.  Could it be refactored to clarify flow?  Consider doing all the checks first to decide what work to do (or error to return), then do the work?

File: fsw/cfe-core/unit-test/tbl_UT.c [hunk 3]
```diff
@@ -2857,7 +2857,7 @@ void Test_CFE_TBL_Load(void)
     RtnCode = CFE_TBL_Load(App1TblHandle1,
                            CFE_TBL_SRC_FILE,
                            "TblSrcFileName.dat");
-    EventsCorrect = (UT_EventIsInHistory(CFE_TBL_LOAD_ERR_EID) == true &&
+    EventsCorrect = (UT_EventIsInHistory(CFE_SUCCESS) == true &&
```
[the remark this answers is not in the document]
Comment (CDKnightNASA) [reply fsw/cfe-core/unit-test/tbl_UT.c]: a little surprised UT_EventIsInHistory(CFE_SUCCESS) returned true...I replaced CFE_TBL_LOAD_ERR_EID and was using CFE_SUCCESS just to get the compiler to be happy, but forgot to change this check.

File: fsw/cfe-core/src/tbl/cfe_tbl_api.c (hunk 2 above)
Comment (CDKnightNASA) [reply fsw/cfe-core/src/tbl/cfe_tbl_api.c:709]: There's quite a lot of conditional logic (whether loading the table from a file or memory block, whether the load is partial, etc) so can't do a lot "up-front". I've refactored to do fail=>event+return to reduce deeply-nested if struct of original code. Unsure if it could be further refactored without changing/removing behavior.

**Comment by astrogeco** [comment]
CCB 20200422 - Approve and open a new ticket to refactor and reduce complexity. There's lots of return statements and we probably don't want to add more.

**Comment by skliper** [comment]
#600 is the table load refactor, I'll hijack and include the complexity reduction effort

**Comment by jphickey** [comment]
This is breaking the integration-candidate build for me, it is using "ThisAppId" variable uninitialized:

```
/home/joe/code/cfecfs/github/cfe/fsw/cfe-core/src/tbl/cfe_tbl_api.c: In function ‘CFE_TBL_Load’:
/home/joe/code/cfecfs/github/cfe/fsw/cfe-core/src/tbl/cfe_tbl_api.c:687:5: error: ‘ThisAppId’ is used uninitialized in this function [-Werror=uninitialized]
  687 |     CFE_ES_GetAppName(AppName, ThisAppId, OS_MAX_API_NAME);
```

**Comment by astrogeco** [comment]
@CDKnightNASA

File: fsw/cfe-core/src/inc/cfe_tbl_events.h [hunk 4]
```diff
@@ -1052,6 +1041,44 @@
...
+**  \par Type: ERROR
+**
+**  \par Cause:
+**
+**  This event message is generated when an application attempts to load a table already
+**  in progress. Likely due to a race condition.
+**/
+#define CFE_TBL_LOAD_IN_PROGRESS_ERR_EID       100
+
+/** TODO: document */
```
[the remark this answers is not in the document]
Comment (skliper) [reply fsw/cfe-core/src/inc/cfe_tbl_events.h:1067]: Something broke down in our process, these TODO's got into the integration candidate @astrogeco @CDKnightNASA
